concat_yaml.py

```diff
@@ -0,0 +1,321 @@
+    cscs = [item for item in cscs if item[keys[0]][keys[1]] in cscs_specs]
+
+    return c2sm + cscs
+
+
+def remove_duplicate_packages(c2sm, cscs, external):
+    c2sm_package_names = dictkeys_as_set(c2sm)
+    cscs_package_names = dictkeys_as_set(cscs)
+    external_package_names = dictkeys_as_set(external)
+
+    duplicates = (c2sm_package_names & cscs_package_names)
+    for dupl in duplicates:
+        cscs_package_names.remove(dupl)
+
+    duplicates = (c2sm_package_names & external_package_names)
+    for dupl in duplicates:
+        external_package_names.remove(dupl)
+
+    c2sm = remove_from_dict(c2sm, c2sm_package_names)
+    cscs = remove_from_dict(cscs, cscs_package_names)
+    external = remove_from_dict(external, external_package_names)
+
+    c2sm.update(cscs)
+    c2sm.update(external)
+    return c2sm
+
+
+def join_compilers(primary, secondary):
+    print('Join compilers')
+
+    primary_compilers = load_from_yaml(primary)
+    secondary_compilers = load_from_yaml(secondary)
+
+    joint = {}
+    joint['compilers'] = remove_duplicate_compilers(
+        primary_compilers['compilers'], secondary_compilers['compilers'],
+        ['compiler', 'spec'])
+
+    return joint
+
+
+def join_packages(primary, secondary, external):
+    print('Join packages')
+    primary_packages = load_from_yaml(primary)['packages']
+    secondary_packages = load_from_yaml(secondary)['packages']
+    external_packages = load_from_yaml(external)['packages']
+
+    primary_package_names = dictkeys_as_set(primary_packages)
+    secondary_package_names = dictkeys_as_set(secondary_packages)
+    external_package_names = dictkeys_as_set(external_packages)
+
+    duplicates = (primary_package_names & secondary_package_names)
+    for dupl in duplicates:
+        secondary_package_names.remove(dupl)
+
+    duplicates = (primary_package_names & external_package_names)
+    for dupl in duplicates:
+        external_package_names.remove(dupl)
+
+    primary = remove_from_dict(primary_packages, primary_package_names)
+    secondary = remove_from_dict(secondary_packages, secondary_package_names)
+    external = remove_from_dict(external_packages, external_package_names)
+
+    primary.update(secondary)
+    primary.update(external)
+    dict = {}
+    dict['packages'] = primary
+
+    return dict
+
+
+# HELPERS
+
+
+def load_from_yaml(file):
+    print(f'Load yaml file: {file}')
+    with open(file, "r") as f:
+        try:
+            data = yaml.load(f)
+        except yaml.error.MarkedYAMLError as e:
+            raise syaml.SpackYAMLError("error parsing YAML spec:", str(e))
+    return data
+
+
+def specs_from_list_with_keys(spec_list, key_1, key_2):
+    specs = set()
+    for item in spec_list:
+        specs.add(item[key_1][key_2])
+
+    return specs
+
+
+def dictkeys_as_set(dict):
+    keys = set()
+    for spec in dict.keys():
+        keys.add(spec)
+    return keys
+
+
+def remove_from_dict(dict, filter):
```

Comment:
This name tricked me hard!
I assumed it removes the filter, when it actually does the exact opposite.

Reply:
Perfect place for a dict comprehension
`return { k:v for k,v in dict.items() if k in filter }`